src/metabase/query_processor/middleware/parameters/mbql.clj

```diff
@@ -21,21 +23,33 @@
     (Double/parseDouble s)
     (Long/parseLong s)))
 
+(defn- field-type
+  [field-clause]
+  (mbql.u/match-one field-clause
+    [:field (id :guard integer?) _]  ((some-fn :effective-type :base-type)
+                                      (lib.metadata.protocols/field (qp.store/metadata-provider) id))
+    [:field (_ :guard string?) opts] (:base-type opts)))
+
+(defn- expression-type
+  [query expression-clause]
+  (mbql.u/match-one expression-clause
+    [:expression (expression-name :guard string?)]
+    (lib/type-of (lib/query (qp.store/metadata-provider) (lib.convert/->pMBQL query))
+                 (lib.convert/->pMBQL &match))))
```

Comment:
This might get too expensive when there are many parameters bound to expressions, but I'm not sure if it's worth optimizing right now.